openquake/hazardlib/shakemap/parsers.py

```diff
@@ -205,6 +209,89 @@ def get_array_usgs_xml(kind, grid_url, uncertainty_url=None):
+    """
+    ftype = rup_json['features'][0]['geometry']['type']
+    assert ftype == 'MultiPolygon', ftype
+    multicoords = rup_json['features'][0]['geometry']['coordinates'][0]
+    hyp_depth = rup_json['metadata']['depth']
+    rake = rup_json['metadata'].get('rake', 0)
+    trt = 'Active Shallow Crust' if hyp_depth < 50 else 'Subduction IntraSlab'
+    Mw = rup_json['metadata']['mag']
+    rup = get_multiplanar(multicoords, Mw, rake, trt)
+    return rup
+
+
+# Namespace
+F64 = numpy.float64
+NAMESPACE = 'http://openquake.org/xmlns/nrml/0.4'
+NRML05 = 'http://openquake.org/xmlns/nrml/0.5'
+GML_NAMESPACE = 'http://www.opengis.net/gml'
+SERIALIZE_NS_MAP = {None: NAMESPACE, 'gml': GML_NAMESPACE}
+PARSE_NS_MAP = {'nrml': NAMESPACE, 'gml': GML_NAMESPACE}
+
+
+# Write node
+def nwrite(nodes, output=sys.stdout, fmt='%.5f', gml=True, xmlns=None):
+    """
+    Convert nodes into a NRML file. output must be a file
+    object open in write mode. If you want to perform a
+    consistency check, open it in read-write mode, then it will
+    be read after creation and validated.
+    :params nodes: an iterable over Node objects
+    :params output: a file-like object in write or read-write mode
+    :param fmt: format used for writing the floats (default '%.7E')
+    :param gml: add the http://www.opengis.net/gml namespace
+    :param xmlns: NRML namespace like http://openquake.org/xmlns/nrml/0.4
+    """
+    root = Node('nrml', nodes=nodes)
+    namespaces = {xmlns or NRML05: ''}
+    if gml:
+        namespaces[GML_NAMESPACE] = 'gml:'
+    with floatformat(fmt):
+        node_to_xml(root, output, namespaces)
+    # if hasattr(output, 'mode') and '+' in output.mode:  # read-write mode
+    #     output.seek(0)
+    #     read(output)  # validate the written file
+
+
+# Convert rupture to file
+def rup_to_file(rup, outfile, commentstr):
+    # Determine geometry
+    geom = rup.surface.surface_nodes[0].tag
+    name = ""
+    if len(rup.surface.surface_nodes) > 1:
+        name = 'multiPlanesRupture'
+    elif geom == 'planarSurface':
+        name = 'singlePlaneRupture'
+    elif geom == 'simpleFaultGeometry':
+        name = 'simpleFaultRupture'
+    elif geom == 'complexFaultGeometry':
+        name = 'complexFaultRupture'
+    elif geom == 'griddedSurface':
+        name = 'griddedRupture'
+    elif geom == 'kiteSurface':
+        name = 'kiteSurface'
+    # Arrange node
+    h = rup.hypocenter
+    hp_dict = dict(lon=h.longitude, lat=h.latitude, depth=h.depth)
+    geom_nodes = [Node('magnitude', {}, rup.mag),
+                  Node('rake', {}, rup.rake),
+                  Node('hypocenter', hp_dict)]
+    geom_nodes.extend(rup.surface.surface_nodes)
+    rupt_nodes = [Node(name, nodes=geom_nodes)]
+    node = Node('nrml', nodes=rupt_nodes)
+    # Write file
+    with open(outfile, 'w') as ff:
+        bs = io.BytesIO()
+        nwrite(node, output=bs)
+        ff.write(bs.getvalue().decode("utf-8"))
+        ff.write(commentstr)
```

Comment:
Why not directly using `nrml.write`? What's the purpose of the commentstr?

Reply:
The commentstr is something like: `<!-- Rupture XML automatically generated from USGS ({md['id']}).  Reference: {md['reference']}.-->\n`

Reply:
It would be better to extend `nrml.write` and pass to it the `commentstr` (default empty string), then you could just write
```python
nrml.write(node, ff, commentstr)
```
BTW, the comment would look better at the beginning of the generated file, not at the end.